### ISSUE TITLE
[BUG] pf.run(..., stream=True) doesn't stream output  

### ISSUE THREAD

**Issue by faaany** [microsoft/promptflow#1218]
**Describe the problem**
When using the following code to run my promptflow application in batch mode, I always get the log in one shot rather than see the streamed output. 
```python
from promptflow import PFClient

pf = PFClient()
column_mapping = {
        "query_id": "${data.query_id}",
        "query": "${data.query}",
        "text_passages": "${data.context}"
    }
run_batch = pf.run(flow=flow_path, data=data_path, column_mapping=column_mapping, stream=True)
```
When looking at the source code, I see that the problem lies in [this line of code ](https://github.com/microsoft/promptflow/blob/main/src/promptflow/promptflow/_sdk/operations/_run_operations.py#L95). The `RubSubmitter` doesn't return the created_run immediately. Maybe the `stream` is not intended to return the logs in streaming mode, rather the output of the chatbot? 

If this is the case, is it possible to see the intermediate results when running promptflow in batch mode?

**Comment by D-W-** [comment]
Hi @faaany , yes it's a known issue since local batch executing is synced. 
We have a feature to make batch run async and then the behavior will become:
- `stream is False`: execute run in async, return after started to run.
- `stream is True`: pull logs and wait run complete.

**Comment by faaany** [comment]
> Hi @faaany , yes it's a known issue since local batch executing is synced. We have a feature to make batch run async and then the behavior will become:
> 
> * `stream is False`: execute run in async, return after started to run.
> * `stream is True`: pull logs and wait run complete.

will it also explain my observations below?

- I have a jsonl file with 2 data points and the total execution time of the flow is 35.5s. Looking at the log in `run_info`, I see that the flow of each data point took [15.9, 21.72], which is 37.62 in total. 
- I would expect that the total execution time of the flow should lie around 21.72, rather 37.63..

**Comment by D-W-** [comment]
Hi @faaany , when executing in batch, the data will be processed in parallel, so the total execution time may be less than sum of all data points' execution time.

**Comment by faaany** [comment]
> Hi @faaany , may I know where do you get the 35.5s of total execution time? And do you want to get actual flow execution time or do you want to understand what happened between the gap of 21.72s and 35.5s?

I measured the total execution time like this:
```python
start = time.time()
run_batch = pf.run(flow=flow_path, data=data_path, column_mapping=column_mapping, stream=True, environment_variables={"PF_WORKER_COUNT": pf_worker_count})
end = time.time()
```

I want to know why the total execution time is 35.5s, not around 21.72s. If I understood it correctly, for a dataset with 2 samples, there will be 2 threads running each sample in parallel. So if one data point takes 15.9s and the other takes 21.72s, I would expect that the total execution time of the flow will be around 21.72s, rather 35.5s, which is almost the sum of 15.9s and 21.72s.

Using another example, I have 2 jsonl files, one with 100 samples and the other with 200 samples. What I have observed is that the total execution time of the flow for the 200 samples almost doubles the time for 100 samples.

**Comment by Hhhilulu** [comment]
Hi, @faaany 
The default worker_count concurrency is 16. And seems like you changed it by environment variable.
So, you need to find the actual concurrency by searching "process count:" in your log.

**Ideally, it needs to be larger than 200 to avoid doubling.**

start = time.time()
run_batch = pf.run(flow=flow_path, data=data_path, column_mapping=column_mapping, stream=True, **_environment_variables={"PF_WORKER_COUNT": pf_worker_count_**})
end = time.time()
 
We have some sophisticate logic to set process count logic, you can found in [_line_execution_process_pool.py](https://github.com/microsoft/promptflow/blob/e4e7b71f2003d7fda0a64991628631dd9d24a1b2/src/promptflow/promptflow/executor/_line_execution_process_pool.py#L199C9-L205C78)

Since different startup methods of different systems have relatively large differences in memory management, we have different ways to calculate process count according to different systems and different process startup methods. The default process startup method of Linux is fork,  windows/macOS(after the python3.8 version) is spawn 

When using fork, compare **worker_count** with **the number of batch run** **execution lines**, and take the minimum value as the process count.

When using spawn, compare **worker_count**, **the number of batch run** **execution lines** and calculate the **available worker_count based on the current available memory**, and take the minimum value as the process count.

**Comment by faaany** [comment]
> Hi, @faaany The default worker_count concurrency is 16. And seems like you changed it by environment variable. So, you need to find the actual concurrency by searching "process count:" in your log.
> 
> **Ideally, it needs to be larger than 200 to avoid doubling.**
> 
> start = time.time() run_batch = pf.run(flow=flow_path, data=data_path, column_mapping=column_mapping, stream=True, **_environment_variables={"PF_WORKER_COUNT": pf_worker_count_**}) end = time.time()
> 
> We have some sophisticate logic to set process count logic, you can found in [_line_execution_process_pool.py](https://github.com/microsoft/promptflow/blob/e4e7b71f2003d7fda0a64991628631dd9d24a1b2/src/promptflow/promptflow/executor/_line_execution_process_pool.py#L199C9-L205C78)
> 
> Since different startup methods of different systems have relatively large differences in memory management, we have different ways to calculate process count according to different systems and different process startup methods. The default process startup method of Linux is fork, windows/macOS(after the python3.8 version) is spawn
> 
> When using fork, compare **worker_count** with **the number of batch run** **execution lines**, and take the minimum value as the process count.
> 
> When using spawn, compare **worker_count**, **the number of batch run** **execution lines** and calculate the **available worker_count based on the current available memory**, and take the minimum value as the process count.

Hi @Hhhilulu , thanks for the detailed explanation. But it doesn't really answer my question. Let me ask in another way:

have you guys ever tested the batch run performance? E.g. by setting pf_worker_count to 200 and measure the execution time of run for one jsonl file with 100 samples and one jsonl file with 50 samples? Theoretically, the execution time for these 2 cases should be equal, am I right?

**Comment by Hhhilulu** [comment]
Hi, @faaany, thanks for reaching out.
We conducted performance testing on Linux environment and the overhead was **_under five seconds_**, inrrelavent to the input volume. 
Specifically, the test setup is to trigger a batch run on a flow with only one node that sleeps for 10 seconds, we gathered statistics for 60 parallelism on 30,60,120,1200 input lines and 100 parallelism on 50,100,1000 input lines, and here's the result:
 
Here is the durations when the worker count is 60:
| **Total lines** | **1200** | **120** | **60** | **30** |
|-|-|-|-|-|  
| Durations |  203.6 | 22.8 | 12.6 | 12.4 |
 
Here is the durations when the worker count is 100:
| **Total lines** | **1000** | **100** | **50** |
|-|-|-|-| 
| Durations | 104.3 | 12.7 | 12.5 |
 
We don't see general perf issue in our test, could you please provide more info about your flow, so that we can help to take a look on the case? Thanks.

**Comment by faaany** [comment]
> Hi, @faaany, thanks for reaching out. We conducted performance testing on Linux environment and the overhead was **_under five seconds_**, inrrelavent to the input volume. Specifically, the test setup is to trigger a batch run on a flow with only one node that sleeps for 10 seconds, we gathered statistics for 60 parallelism on 30,60,120,1200 input lines and 100 parallelism on 50,100,1000 input lines, and here's the result:
> 
> Here is the durations when the worker count is 60:
> 
> **Total lines**	**1200**	**120**	**60**	**30**
> Durations	203.6	22.8	12.6	12.4
> Here is the durations when the worker count is 100:
> 
> **Total lines**	**1000**	**100**	**50**
> Durations	104.3	12.7	12.5
> We don't see general perf issue in our test, could you please provide more info about your flow, so that we can help to take a look on the case? Thanks.

thanks so much for the data! it is good to see that the scaling performance is fine.
May I know which promptflow version do you use?

**Comment by Hhhilulu** [comment]
Hi, @faaany 
The latest version. And the parallelism has not changed for several months.

**Comment by ybz21** [comment]
> Hi @faaany , yes it's a known issue since local batch executing is synced. We have a feature to make batch run async and then the behavior will become:
> 
> * `stream is False`: execute run in async, return after started to run.
> * `stream is True`: pull logs and wait run complete.

when will this feature be released?

**Comment by faaany** [comment]
> Hi, @faaany The latest version. And the parallelism has not changed for several months.

I also updated my version to the latest 1.1.1. But the performance still couldn't meet my expectations. Can we set up a quick call for discussion? It is a bit hard to replicate my use case, since I have all my services locally. You can find my E-mail address on my GitHub profile. Many Thanks!

**Comment by D-W-** [comment]
> > Hi @faaany , yes it's a known issue since local batch executing is synced. We have a feature to make batch run async and then the behavior will become:
> > 
> > * `stream is False`: execute run in async, return after started to run.
> > * `stream is True`: pull logs and wait run complete.
> 
> when will this feature be released?

Hi @ybz21 , we plan to release this in this month, probably will be 1.2.0 or 1.3.0. I'll update here after it released.

**Comment by D-W-** [comment]
> > Hi, @faaany The latest version. And the parallelism has not changed for several months.
> 
> I updated my version to the latest 1.1.1. But the performance still couldn't meet my expectations. Can we set up a quick call for discussion? It is a bit hard to reproduce my use case since I have all my services locally. You can find my E-mail address on my GitHub profile. Many Thanks!

Hi @faaany , the above result are from a simple test flow(see attachment). 
The perf improvement after increasing worker count maybe related to actual code logic.
For example, if the code is compute intensive or I/O intensive. The speed will stop increase after reaching the upper limit of CPU or IO.
So in summary, it's not guaranteed to increase speed by 2x by setting worker count to 2x.

We have 2 ways to keep investigate the issue:

1. Use `pf run visualize` and compare `Api calls` for different worker count. Reference [here](https://microsoft.github.io/promptflow/how-to-guides/run-and-evaluate-a-flow/index.html) for details. 
2. Use `cprofile` to profile the program. Reference [here](https://docs.python.org/3/library/profile.html) for details. If the result shows promptflow APIs took long time to run. We can help check if there's improvement we can do.

For follow up perf related questions, could you create a separate issue to track? We'll use this issue to track the `stream` thing.

[sample-flow.zip](https://github.com/microsoft/promptflow/files/13577605/sample-flow.zip)

**Comment by faaany** [comment]
> > > Hi, @faaany The latest version. And the parallelism has not changed for several months.
> > 
> > 
> > I updated my version to the latest 1.1.1. But the performance still couldn't meet my expectations. Can we set up a quick call for discussion? It is a bit hard to reproduce my use case since I have all my services locally. You can find my E-mail address on my GitHub profile. Many Thanks!
> 
> Hi @faaany , the above result are from a simple test flow(see attachment). The perf improvement after increasing worker count maybe related to actual code logic. For example, if the code is compute intensive or I/O intensive. The speed will stop increase after reaching the upper limit of CPU or IO. So in summary, it's not guaranteed to increase speed by 2x by setting worker count to 2x.
> 
> We have 2 ways to keep investigate the issue:
> 
> 1. Use `pf run visualize` and compare `Api calls` for different worker count. Reference [here](https://microsoft.github.io/promptflow/how-to-guides/run-and-evaluate-a-flow/index.html) for details.
> 2. Use `cprofile` to profile the program. Reference [here](https://docs.python.org/3/library/profile.html) for details. If the result shows promptflow APIs took long time to run. We can help check if there's improvement we can do.
> 
> For follow up perf related questions, could you create a separate issue to track? We'll use this issue to track the `stream` thing.
> 
> [sample-flow.zip](https://github.com/microsoft/promptflow/files/13577605/sample-flow.zip)

Sure, we can close this issue and let me first try the methods you suggested to investigate the performance issue I have. And if there is indeed a performance issue, I will open another thread for that. Thanks a lot!

**Comment by D-W-** [comment]
> > > Hi @faaany , yes it's a known issue since local batch executing is synced. We have a feature to make batch run async and then the behavior will become:
> > > 
> > > * `stream is False`: execute run in async, return after started to run.
> > > * `stream is True`: pull logs and wait run complete.
> > 
> > 
> > when will this feature be released?
> 
> Hi @ybz21 , we plan to release this in this month, probably will be 1.2.0 or 1.3.0. I'll update here after it released.

Hi @ybz21 , due to plan change. We'll support flow run streaming in next months release. I'll update here once released.
